### ISSUE TITLE
CMS Enabled / Requests for MicrosoftAjax.de.js / Multilanguage

### ISSUE THREAD

**Issue by Andy9999** [dnnsoftware/Dnn.Platform#4559]
We noticed that our 9.4.1 DNN makes requests to MicrosoftAjax.de.js if the locale is set to German and CMS is enabled. While our page doesn't require Ajax and works with Java Script disabled, it adds a request that fails during page load, because localized versions of MicrosoftAjax.js do not exist afaik. It resulted in some flickering when switching between pages.

The fix was to change this in AJAX.cs:
using (var scriptManager = new ScriptManager //RadScriptManager
{
   ID = "ScriptManager",
   EnableScriptGlobalization = true,
   SupportsPartialRendering = true
 })
 
 into that:
using (var scriptManager = new ScriptManager //RadScriptManager
{
   ID = "ScriptManager",
   EnableScriptGlobalization = true,
   SupportsPartialRendering = true,
   EnableScriptLocalization = false
})

**Comment by mitchelsellers** [comment]
Is this still an issue in 9.9.0?

**Comment by Andy9999** [comment]
Yes, I just checked the most up-to-date Ajax.cs version. But I haven't tried it yet on a fresh DNN install.

**Comment by Andy9999** [comment]
I would have a patch for this issue that I could submit in case anyone is interested in this.

Can anyone, running multiple languages, using modules that include Ajax and having CDN enabled for Ajax, confirm that they have the same problem? You should be able to see the failed request, due to the wrong file name generated having the language code added, in a Browser network trace.

I would be glad if I could submit this so I don't have to use a custom build anymore.

Best regards & happy DNNing
Andy9999

**Comment by Andy9999** [comment]
I will test it one more time as a lot of time has passed, but if I remember correctly it also would occur if CDN is disabled. The fix would then be, to simply copy the existing original version and give the copy the expected file name. But I could be wrong. The problem with CDN being enabled is clearly that you cannot create that additional copy on Microsoft's CDN server.

Best regards & happy DNNing
Andy9999

**Comment by bdukes** [comment]
Well, if it doesn't work at all, then we can just always set `EnableScriptLocalization` to `false`, right?

**Comment by Andy9999** [comment]
True. :)

**Comment by Andy9999** [comment]
A tiny update.

This is the URL to Ajax being used: https://ajax.aspnetcdn.com/ajax/4.6/1/MicrosoftAjax.js

There is a folder below with the name globalization and inside there is for example a file "fr-FR.js" or "de-DE.js". I am not sure if this is loaded automatically by the main script.

So this is for example a valid URL: https://ajax.aspnetcdn.com/ajax/4.6/1/globalization/fr-FR.js or https://ajax.aspnetcdn.com/ajax/4.6/1/globalization/de-DE.js

I have opened an issue over here and hope someone can clarify: https://github.com/dotnet/AspNetDocs/issues/594

Best regards & happy DNNing
Andy9999

**Comment by Andy9999** [comment]
Here is some interesting documentation about globalization support in ASP.NET AJAX applications: https://docs.microsoft.com/en-us/archive/msdn-magazine/2008/january/globalization-support-in-asp-net-ajax-applications

According to that article, setting EnableScriptLocalization  to true (the default) should cause the following to be included (which resembles of the fr-FR.js content for example):
var __cultureInfo = '{"name":"en-GB",
  "numberFormat":{ "CurrencyDecimalDigits":2, "CurrencyDecimalSeparator":".","IsReadOnly":true, 
  "CurrencyGroupSizes":[3],"NumberGroupSizes":[3], "PercentGroupSizes":[3],"CurrencyGroupSeparator":", ",
  "CurrencySymbol":"£","NaNSymbol":"NaN", "CurrencyNegativePattern":1,"NumberNegativePattern":1, 
  "PercentPositivePattern":0,"PercentNegativePattern":0, "NegativeInfinitySymbol":"-Infinity","NegativeSign":"- ...
  
Maybe this problem occurs in conjunction with EnableMsAjaxCDN being enabled only.

Here is one other person complaining, in this case about MicrosoftAjax.zh-Hant.js:
https://github.com/dotnet/AspNetCore.Docs/issues/10133

Am still wondering what is going on there and how the js files inside of that globalization folder should get loaded.

Best regards & happy DNNing
Andy9999

**Comment by Andy9999** [comment]
LOL. Case closed -> https://github.com/dotnet/AspNetDocs/issues/594

**Comment by Andy9999** [comment]
Maybe referring me to StackOverflow was helpful:
https://stackoverflow.com/questions/817046/what-about-script-globalization-of-microsoftajax-js-in-asp-net-mvc

Documenting it properly would have been helpful in the first place. Looks like a bug to me. But anyways. At least something to work with.

Best regards & happy DNNing
Andy9999

**Comment by Andy9999** [comment]
Seems like this is still an issue.

Seems like the CDN has sub folders for each language instead of a custom file name.

This only occurs if a language different than English is used and CDN is enabled. I saw 404s when requesting from the MS CDN. 

If CDN is on, it does not make sense to have EnableScriptLocalization set to TRUE (the default value) if the files are not in the expected folder.

Will provide more info tomorrow. I just resurrected a 14 year old stackoverflow thread: https://stackoverflow.com/questions/5885974/localized-microsoft-ajax-cdn-reference-not-working-in-asp-net-webforms

Currently I set localization hard coded to false. But maybe ScriptManager is smarter now and somehow finding the localized Ajax files on their CDN (if MS ever uploaded the missing files).

**Comment by Andy9999** [comment]
I did search a bit more and found that this URL works: https://ajax.aspnetcdn.com/ajax/4.0/1/globalization/DE-de.js

It could be loaded in addition to the MicrosoftAjax.js.

But that would be two HTTP requests then.

I think in our case, we already have another CDN for web traffic, we saw a performance increase if letting Ajax be loaded from the MS CDN leading to faster page load times.

Would it be acceptable to enforce the non-localized version if CDN is enabled for Ajax?

If someone could let me know how I can get the information if AjaxCDN is on or off the best way I can change my static implementation and make a pull request. Getting it from the DB table is probably not the right way.

That would at least not 404 anymore.

Or we could try to load both files (the main file and the localization part), but that would be sub optimal performance wise, especially if one opts for CDN.

**Comment by Andy9999** [comment]
I just tested on a localized site that has CDN enabled and I don't see the MicrosoftAjax.XX.js request anymore. Maybe it was fixed in ScriptManager. So this is not an issue anymore. :)
